From c013e76b1cf56b89b5faf941c8a5441433f355b8 Mon Sep 17 00:00:00 2001
From: Jonh Wendell <jonh.wendell@redhat.com>
Date: Tue, 8 Feb 2022 16:02:21 -0500
Subject: Restore proxyv2 Dockerfile

For now we are still using upstream proxy, so we must use their
Dockerfile. Once we switch to maistra proxy, we can revert this.
---
 pilot/docker/Dockerfile.proxyv2         | 26 ++++++++++++-----
 pilot/docker/Dockerfile.proxyv2.maistra | 39 +++++++++++++++++++++++++
 2 files changed, 57 insertions(+), 8 deletions(-)
 create mode 100644 pilot/docker/Dockerfile.proxyv2.maistra

diff --git a/pilot/docker/Dockerfile.proxyv2 b/pilot/docker/Dockerfile.proxyv2
index 7680d71b74..02cded141b 100644
--- a/pilot/docker/Dockerfile.proxyv2
+++ b/pilot/docker/Dockerfile.proxyv2
@@ -1,9 +1,22 @@
-FROM centos:8
+# BASE_DISTRIBUTION is used to switch between the old base distribution and distroless base images
+ARG BASE_DISTRIBUTION=debug
 
-# hadolint ignore=DL3039,DL3041
-RUN dnf -y upgrade --refresh && \
-    dnf -y install iptables iproute openssl && \
-    dnf -y clean all
+# Version is the base image version from the TLD Makefile
+ARG BASE_VERSION=latest
+
+# The following section is used as base image if BASE_DISTRIBUTION=debug
+FROM gcr.io/istio-release/base:${BASE_VERSION} as debug
+
+# The following section is used as base image if BASE_DISTRIBUTION=distroless
+# This image is a custom built debian11 distroless image with multiarchitecture support.
+# It is built on the base distroless image, with iptables binary and libraries added
+# The source can be found at https://github.com/istio/distroless/tree/iptables
+# This version is from commit 20b20ec36f4621830ff0bfdec519577074df47ca.
+FROM gcr.io/istio-release/iptables@sha256:bae9287d64be13179b7bc794ec3db26bd5c5fe3fb591c484992366314c9a7d3d as distroless
+
+# This will build the final image based on either debug or distroless from above
+# hadolint ignore=DL3006
+FROM ${BASE_DISTRIBUTION:-debug}
 
 WORKDIR /
 
@@ -15,9 +28,6 @@ ARG SIDECAR=envoy
 COPY envoy_bootstrap.json /var/lib/istio/envoy/envoy_bootstrap_tmpl.json
 COPY gcp_envoy_bootstrap.json /var/lib/istio/envoy/gcp_envoy_bootstrap_tmpl.json
 
-RUN useradd -m --uid 1337 istio-proxy
-RUN chown -R istio-proxy /var/lib/istio
-
 # Install Envoy.
 ARG TARGETARCH
 COPY ${TARGETARCH:-amd64}/${SIDECAR} /usr/local/bin/${SIDECAR}
diff --git a/pilot/docker/Dockerfile.proxyv2.maistra b/pilot/docker/Dockerfile.proxyv2.maistra
new file mode 100644
index 0000000000..7680d71b74
--- /dev/null
+++ b/pilot/docker/Dockerfile.proxyv2.maistra
@@ -0,0 +1,39 @@
+FROM centos:8
+
+# hadolint ignore=DL3039,DL3041
+RUN dnf -y upgrade --refresh && \
+    dnf -y install iptables iproute openssl && \
+    dnf -y clean all
+
+WORKDIR /
+
+ARG proxy_version
+ARG istio_version
+ARG SIDECAR=envoy
+
+# Copy Envoy bootstrap templates used by pilot-agent
+COPY envoy_bootstrap.json /var/lib/istio/envoy/envoy_bootstrap_tmpl.json
+COPY gcp_envoy_bootstrap.json /var/lib/istio/envoy/gcp_envoy_bootstrap_tmpl.json
+
+RUN useradd -m --uid 1337 istio-proxy
+RUN chown -R istio-proxy /var/lib/istio
+
+# Install Envoy.
+ARG TARGETARCH
+COPY ${TARGETARCH:-amd64}/${SIDECAR} /usr/local/bin/${SIDECAR}
+
+# Environment variable indicating the exact proxy sha - for debugging or version-specific configs 
+ENV ISTIO_META_ISTIO_PROXY_SHA $proxy_version
+# Environment variable indicating the exact build, for debugging
+ENV ISTIO_META_ISTIO_VERSION $istio_version
+
+ARG TARGETARCH
+COPY ${TARGETARCH:-amd64}/pilot-agent /usr/local/bin/pilot-agent
+
+COPY stats-filter.wasm /etc/istio/extensions/stats-filter.wasm
+COPY stats-filter.compiled.wasm /etc/istio/extensions/stats-filter.compiled.wasm
+COPY metadata-exchange-filter.wasm /etc/istio/extensions/metadata-exchange-filter.wasm
+COPY metadata-exchange-filter.compiled.wasm /etc/istio/extensions/metadata-exchange-filter.compiled.wasm
+
+# The pilot-agent will bootstrap Envoy.
+ENTRYPOINT ["/usr/local/bin/pilot-agent"]
-- 
2.35.3

